include/derecho/core/detail/external_group_impl.hpp

```diff
@@ -0,0 +1,436 @@
+#endif
+
+    get_view(-1);
+
+    uint64_t view_max_rpc_reply_payload_size = 0;
+    uint32_t view_max_rpc_window_size = 0;
+
+    for(subgroup_id_t subgroup_id = 0; subgroup_id < curr_view->subgroup_shard_views.size(); ++subgroup_id) {
+        uint64_t max_payload_size = 0;
+        uint32_t num_shards = curr_view->subgroup_shard_views.at(subgroup_id).size();
+        for(uint32_t shard_num = 0; shard_num < num_shards; ++shard_num) {
+            SubView& shard_view = curr_view->subgroup_shard_views.at(subgroup_id).at(shard_num);
+            const DerechoParams& profile = DerechoParams::from_profile(shard_view.profile);
+            uint64_t payload_size = profile.max_msg_size - sizeof(header);
+            max_payload_size = std::max(payload_size, max_payload_size);
+            view_max_rpc_reply_payload_size = std::max(
+                    profile.max_reply_msg_size - sizeof(header),
+                    view_max_rpc_reply_payload_size);
+            view_max_rpc_window_size = std::max(profile.window_size, view_max_rpc_window_size);
+        }
+        max_payload_sizes[subgroup_id] = max_payload_size;
+    }
+
+    p2p_connections = std::make_unique<sst::P2PConnectionManager>(sst::P2PParams{
+            my_id,
+            getConfUInt32(CONF_DERECHO_P2P_WINDOW_SIZE),
+            view_max_rpc_window_size,
+            getConfUInt64(CONF_DERECHO_MAX_P2P_REPLY_PAYLOAD_SIZE) + sizeof(header),
+            getConfUInt64(CONF_DERECHO_MAX_P2P_REQUEST_PAYLOAD_SIZE) + sizeof(header),
+            view_max_rpc_reply_payload_size + sizeof(header),
+            true,
+            NULL});
+
+    rpc_thread = std::thread(&ExternalGroup<ReplicatedTypes...>::p2p_receive_loop, this);
+}  // namespace derecho
+
+template <typename... ReplicatedTypes>
+ExternalGroup<ReplicatedTypes...>::~ExternalGroup() {
+    thread_shutdown = true;
+    if(rpc_thread.joinable()) {
+        rpc_thread.join();
+    }
+}
+
+template <typename... ReplicatedTypes>
+bool ExternalGroup<ReplicatedTypes...>::get_view(const node_id_t nid) {
+    try {
+        tcp::socket sock = curr_view ? 
+        tcp::socket(std::get<0>(curr_view->member_ips_and_ports[curr_view->rank_of(nid)]),
+                    std::get<PORT_TYPE::GMS>(curr_view->member_ips_and_ports[curr_view->rank_of(nid)]),
+                    false) :
+        tcp::socket(getConfString(CONF_DERECHO_LEADER_IP), getConfUInt16(CONF_DERECHO_LEADER_GMS_PORT));
+        
+        JoinResponse leader_response;
+        bool success;
+        uint64_t leader_version_hashcode;
+        success = sock.exchange(my_version_hashcode, leader_version_hashcode);
+        if(!success) throw derecho_exception("Failed to exchange version hashcodes with the leader! Leader has crashed.");
+        if(leader_version_hashcode != my_version_hashcode) {
+            throw derecho_exception("Unable to connect to Derecho leader because the leader is running on an incompatible platform or used an incompatible compiler.");
```

Comment:
Why throw an exception here if you're just going to catch it further down in the same function? If you want this function to return false on any error, you could just replace this throw statement with `return false`